pkg/model/multiversion/delta.go

```diff
@@ -0,0 +1,248 @@
+	Spoke *ackmodel.Field
+}
+
+// CRDDelta stores the spec and status deltas for a custom resource.
+type CRDDelta struct {
+	SpecDeltas   []FieldDelta
+	StatusDeltas []FieldDelta
+}
+
+// ComputeCRDFieldsDeltas compares two ackmodel.CRD instances and returns the
+// spec and status fields deltas. spokeCRD is the CRD of the spoke (source) version
+// and hubCRD is the CRD of the hub (destination) version.
+func ComputeCRDFieldsDeltas(spokeCRD, hubCRD *ackmodel.CRD) (*CRDDelta, error) {
+	renames, _, err := hubCRD.GetAllRenames()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get resource field renames: %s", err)
+	}
+
+	specDeltas, err := ComputeFieldsDiff(spokeCRD.SpecFields, hubCRD.SpecFields, renames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute spec fields deltas: %s", err)
+	}
+
+	statusDeltas, err := ComputeFieldsDiff(spokeCRD.StatusFields, hubCRD.StatusFields, renames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute status fields deltas: %s", err)
+	}
+
+	return &CRDDelta{
+		specDeltas,
+		statusDeltas,
+	}, nil
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+// ComputeFieldsDiff computes the difference between two maps of fields. It returns a list
+// of FieldDelta's that contains the the ChangeType and at least one field reference.
+func ComputeFieldsDiff(
+	spokeCRDFields map[string]*ackmodel.Field,
+	hubCRDFields map[string]*ackmodel.Field,
+	// the hub renames
+	renames map[string]string,
+) ([]FieldDelta, error) {
+	// the smallest size of deltas we can get is max(len(spokeCRDFields), len(hubCRDFields))
+	deltas := make([]FieldDelta, 0, max(len(spokeCRDFields), len(hubCRDFields)))
+
+	// collect field names and sort them to ensure a determenistic output order.
+	spokeFieldsNames := []string{}
+	for name := range spokeCRDFields {
+		spokeFieldsNames = append(spokeFieldsNames, name)
+	}
+	sort.Strings(spokeFieldsNames)
+
+	hubFieldsNames := []string{}
+	for name := range hubCRDFields {
+		hubFieldsNames = append(hubFieldsNames, name)
+	}
+	sort.Strings(hubFieldsNames)
+
+	// let's make sure we don't visit fields more than once - especially
+	// when fields are renamed.
+	visitedFields := map[string]*struct{}{}
+
+	// first let's loop over the spokeFieldsNames array and see if we can find
+	// the same field name in hubFieldNames.
+	for _, spokeFieldName := range spokeFieldsNames {
+		spokeField, _ := spokeCRDFields[spokeFieldName]
+		hubField, ok := hubCRDFields[spokeFieldName]
+		// If a field is found in both arrays only three changes are possible:
+		// Intact, TypeChangeChange and ChangeTypeShapeChangedToSecret.
+		// NOTE(a-hilaly): carefull about X -> Y then Z -> X renames. It should
+		// not be allowed.
+		if ok {
+			// mark field as visited.
+			visitedFields[spokeFieldName] = nil
+			// check if field became secret
+			if (spokeField.FieldConfig == nil ||
+				(spokeField.FieldConfig != nil && !spokeField.FieldConfig.IsSecret)) &&
```

Comment:
Do we care about change in other direction for this SecretChangeType? IOW, should there be another if condition where "spokeField" is a secret but "hubField" is not.

Reply:
Good point @vijtrip2 - it's very unlikely to happen but i should add a case for that.